storage/innobase/handler/handler0alter.cc

```diff
@@ -4220,103 +4220,145 @@ innobase_dropping_foreign(
-	bool			drop)
+bool innobase_check_foreigns_nullability(const dict_table_t *user_table,
+                                         dict_foreign_t     **drop_fk,
+                                         ulint              n_drop_fk,
+                                         const char*        col_name,
+                                         uint32_t           new_field_flags)
 {
-	dict_foreign_t*	foreign;
-	ut_ad(mutex_own(&dict_sys.mutex));
-
-	/* Check if any FOREIGN KEY constraints are defined on this
-	column. */
-
-	for (dict_foreign_set::const_iterator it = user_table->foreign_set.begin();
-	     it != user_table->foreign_set.end();
-	     ++it) {
-
-		foreign = *it;
-
-		if (!drop && !(foreign->type
-			       & (DICT_FOREIGN_ON_DELETE_SET_NULL
-				  | DICT_FOREIGN_ON_UPDATE_SET_NULL))) {
-			continue;
-		}
-
-		if (innobase_dropping_foreign(foreign, drop_fk, n_drop_fk)) {
-			continue;
-		}
-
-		for (unsigned f = 0; f < foreign->n_fields; f++) {
-			if (!strcmp(foreign->foreign_col_names[f],
-				    col_name)) {
-				my_error(drop
-					 ? ER_FK_COLUMN_CANNOT_DROP
-					 : ER_FK_COLUMN_NOT_NULL, MYF(0),
-					 col_name, foreign->id);
-				return(true);
-			}
-		}
-	}
+  ut_ad(mutex_own(&dict_sys.mutex));
 
-	if (!drop) {
-		/* SET NULL clauses on foreign key constraints of
-		child tables affect the child tables, not the parent table.
-		The column can be NOT NULL in the parent table. */
-		return(false);
-	}
+  /* Changing from NULL to NOT NULL. So check referenced set */
+  if ((new_field_flags & NOT_NULL_FLAG))
+  {
+    for (dict_foreign_t *foreign : user_table->foreign_set)
+    {
+      if (innobase_dropping_foreign(foreign, drop_fk, n_drop_fk))
+        continue;
 
-	/* Check if any FOREIGN KEY constraints in other tables are
-	referring to the column that is being dropped. */
-	for (dict_foreign_set::const_iterator it
-		= user_table->referenced_set.begin();
-	     it != user_table->referenced_set.end();
-	     ++it) {
+      if ((foreign->type & foreign->UPDATE_CASCADE) &&
+          foreign->check_fk_null_foreign(col_name))
+        goto non_null_error;
 
-		foreign = *it;
+      if (foreign->type & (foreign->DELETE_SET_NULL |
+                           foreign->UPDATE_SET_NULL))
+      {
+        if (foreign->foreign_index
+            && foreign->col_fk_exists(col_name) != UINT_MAX)
+        {
+non_null_error:
+          const char* fid = strchr(foreign->id, '/');
+          fid= fid ? fid + 1 : foreign->id;
+          my_error(ER_FK_COLUMN_NOT_NULL, MYF(0), col_name, fid);
+          return true;
+        }
+      }
+    }
+  }
+  else
+  {
+    for (dict_foreign_t *foreign : user_table->referenced_set)
+    {
+      if ((foreign->type & foreign->UPDATE_CASCADE)
+          && foreign->check_fk_null_referenced(col_name))
+      {
+        char display_name[FN_REFLEN];
+	const int dblen=
+          int(table_name_t(const_cast<char*>(
+                             foreign->foreign_table_name)).dblen());
```

Comment:
In the common formatting style, lines must not be split immediately before a binary operator, such as `&&`, or immediately after `(`.